extensions/basic_auth/plugin.cc

```diff
@@ -0,0 +1,287 @@
+                      auto credential = JsonValueAs<std::string>(credentials);
+                      if (credential.second !=
+                          Wasm::Common::JsonParserResultDetail::OK) {
+                        LOG_WARN("unexpected credential");
+                        return false;
+                      }
+                      rules.encoded_credentials.insert(
+                          Base64::encode(credential.first.value().data(),
+                                         credential.first.value().size()));
+                      return true;
+                    })) {
+              LOG_WARN("Failed to parse configuration for credentials.");
+              return false;
+            }
+
+            if (match == "prefix") {
+              rules.pattern = Prefix;
+            } else if (match == "exact") {
+              rules.pattern = Exact;
+            } else if (match == "suffix") {
+              rules.pattern = Suffix;
+            }
+            rules.request_path = request_path;
+            for (auto& methods : request_methods) {
+              basic_auth_configuration_[methods].push_back(rules);
+            }
+            return true;
+          })) {
+    LOG_WARN("Failed to parse configuration information.");
+  }
+  return true;
+}
+
+FilterHeadersStatus PluginContext::credentialsCheck(
+    const PluginRootContext::BasicAuthConfigRule& rule,
+    const std::string& authorization_header) {
+  auto credential_iter = rule.encoded_credentials.find("");
+  auto credential_size = rule.encoded_credentials.size();
+  // Check if credential set is of of size 1 and its an empty string.
+  if (credential_size == 1 &&
+      credential_iter != rule.encoded_credentials.end()) {
+    deniedNoAccessToCredentials();
+    return FilterHeadersStatus::StopIteration;
+  }
+  // Check if there is an authorization header
+  if (authorization_header.compare("") == 0) {
+    deniedNoAuthorizationHeader();
+    return FilterHeadersStatus::StopIteration;
+  }
+  absl::string_view basic_authorization_header = authorization_header;
+  // Check if the Basic auth header starts with "Basic "
+  if (!absl::ConsumePrefix(&basic_authorization_header, "Basic ")) {
+    deniedNoBasicAuthData();
+    return FilterHeadersStatus::StopIteration;
+  }
+  auto auth_credential_iter =
+      rule.encoded_credentials.find(std::string(basic_authorization_header));
+  // Check if encoded credential is part of the encoded_credentials
+  // set from our container to grant or deny access.
+  if (auth_credential_iter == rule.encoded_credentials.end()) {
+    deniedInvalidCredentials();
+    return FilterHeadersStatus::StopIteration;
+  } else {
+    return FilterHeadersStatus::Continue;
+  }
+  return FilterHeadersStatus::StopIteration;
+}
+
+FilterHeadersStatus PluginContext::onRequestHeaders(uint32_t, bool) {
+  auto basic_auth_configuration = rootContext()->basicAuthConfigurationValue();
+  auto request_path = getRequestHeader(":path")->toString();
```

Comment:
I don't think you will need to get them as string. This will make a copy of the path, which is inefficient. `auto request_path = getRequestHeader(":path")` returns `WasmData` and I think you can use string_view for most of the code beneath: `request_path->view()`

Reply:
Same comment to method and authorization header.